File: 03 - Binary Search/E141.Sqrt(x).py
class Solution:
    """
    @param x: An integer
    @return: The sqrt of x
    """
    def sqrt(self, x):
        if x < 0:
            return 0

        # write your code here
        left, right = 0, x if x < 2 else x // 2
        while left + 1 < right:
            mid = left + (right - left) // 2
            squre = mid * mid
            if squre == x:
                return mid
            elif squre < x:
                left = mid
            else:
                right = mid

        if right * right <= x:
            return right

        return left

File: 03 - Binary Search/E141.Sqrt(x).java
public class Solution {
    /**
     * @param x: An integer
     * @return: The sqrt of x
     */
    public int sqrt(int x) {
        // write your code here
        // method 2
        if (x <= 0) return 0;

        int l = 1, r = x < 2 ? x : x / 2;
        while (l + 1 < r) {
            int mid = l + (r - l) / 2;
            int quotient = x / mid;
            if (quotient == mid) return mid;

            if (quotient > mid) {
                l = mid;
            } else {
                r = mid;
            }
        }

        return x / r >= r ? r : l;

        // method 1
        // if (x <= 0) return 0;
        // if (x < 2) return x;

        // long l = 1, r = x / 2;
        // while (l + 1 < r) {
        //     long mid = l + (r - l) / 2;
        //     long square = mid * mid;
        //     if (square == x) return (int) mid;

        //     if (square < x) {
        //         l = mid;
        //     } else {
        //         r = mid;
        //     }
        // }

        // return (int) (r * r <= x ? r : l);
    }
}
